### PR TITLE
Gitlabbinding

### LINKED ISSUE #1149: Create a GitLabBinding
Create a GitLabBinding as mentioned here:
https://github.com/knative/eventing-contrib/pull/1120#pullrequestreview-395511741

/assign

### PR DESCRIPTION
**Pull request by tzununbekov**
Fixes #1149

Basically, a shameless copy of [GitHubBinding](https://github.com/knative/eventing-contrib/pull/1113)

**Release Note**

```release-note
GitLabBinding introduced
```

### CODE REVIEW

File: gitlab/pkg/apis/bindings/v1alpha1/gitlabbinding_lifecycle.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	}
+	for i := range spec.Containers {
+		spec.Containers[i].VolumeMounts = append(spec.Containers[i].VolumeMounts, volumeMount)
+	}
+}
+
+func (sb *GitLabBinding) Undo(ctx context.Context, ps *duckv1.WithPod) {
+	spec := ps.Spec.Template.Spec
+
+	// Make sure the PodSpec does NOT have the github volume.
```
Comment (lberk) [inline gitlab/pkg/apis/bindings/v1alpha1/gitlabbinding_lifecycle.go:113]: I'd hope it doesn't have the github volume :sweat_smile:

File: gitlab/pkg/apis/bindings/v1alpha1/gitlabbinding_lifecycle.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	for i, v := range spec.Volumes {
+		if v.Name == gitlab.VolumeName {
+			ps.Spec.Template.Spec.Volumes = append(spec.Volumes[:i], spec.Volumes[i+1:]...)
+			break
+		}
+	}
+
+	// Make sure that none of the [init]containers have the github volume mount
```
Comment (lberk) [inline gitlab/pkg/apis/bindings/v1alpha1/gitlabbinding_lifecycle.go:121]: same here :)

File: gitlab/cmd/webhook/main.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (lberk) [inline gitlab/cmd/webhook/main.go:2]: should probably be 2020, right?

File: gitlab/cmd/webhook/main.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+				return ctx, nil
+			},
+			opts...,
+		)
+	}
+}
+
+func main() {
+	ctx := webhook.WithOptions(signals.NewContext(), webhook.Options{
+		ServiceName: "gitlab-source-webhook",
```
Comment (lberk) [inline gitlab/cmd/webhook/main.go:125]: why not `logconfig.WebhookName()` and ensure we match the name passed to `sharedmain.WebhookMainWithContext` ?

Comment (lberk) [reply gitlab/cmd/webhook/main.go:125]: aha, you've named it `gitlab-source-webhook` in the service yaml, gotcha :+1:

File: gitlab/client.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	return string(data), nil
+}
+
+// AccessToken reads the file named accessToken that is mounted by the GitLabBinding.
+func AccessToken() (string, error) {
+	return ReadKey(AccessTokenKey)
+}
+
+// New instantiates a new gitlab client from the access token from the GitLabBinding
+func New(ctx context.Context) (*gitlab.Client, error) {
```
Comment (mattmoor) [inline gitlab/client.go:49]: This is awesome!  One realization I had this morning (and was chatting with @lberk about) was aligning the changes we make to the environment with existing tooling.  e.g. in the case of GitHubBinding, I think we should support the `hub` CLI.  Not sure if there is a Gitlab equivalent, or if it's auth is flexible enough to just trivially add this, but here's the Github issue: https://github.com/knative/eventing-contrib/issues/1185

File: gitlab/pkg/apis/bindings/v1alpha1/gitlabbinding_lifecycle.go (hunk 1 above)
Comment (tzununbekov) [reply gitlab/pkg/apis/bindings/v1alpha1/gitlabbinding_lifecycle.go]: :see_no_evil:

**Comment by lberk** [comment]
/lgtm
/approve
(I'm not sure what's going on with e2e testing atm)

**Pull request merged**